### ISSUE TITLE
[TechDocs] MyST and/or Sphinx support

### ISSUE THREAD

**Issue by ksdaftari** [backstage/backstage#8642]
I know in documentation say currently just support markdown as built on MkDocs, but wondering if possible plans for MyST support (https://myst-parser.readthedocs.io/en/latest/) and perhaps more generally sphinx support in future. perhaps this is out of scope of this project, but wanted to ask just in case.

**Comment by soapraj** [comment]
@ksdaftari Hi! At this point there is no plans to support alternative static site generators in the project.

**Comment by garyniemen** [comment]
Hi @ksdaftari - there are no plans at present to support other source formats such as RST. One of the founding principles of TechDocs was to fiercely optimise for engineers so we opted to keep the source format as easy to use as possible - hence markdown. We have had discussions in the team here around supporting MDX and alternative static site generators - but these items are not on our roadmap at present. I just updated our roadmap on backstage.io with updated information (but the page is not merged yet).

**Comment by ksdaftari** [comment]
@garyniemen understood,I can understand that principle and definitely agree more enforcement to markdown is likely best. Would have been nice to have MyST which is extension of markdown but can understand perhaps even if wanted to not something this early on would want to implement.

Curious I think one of the main features of MyST that is nice is its ability it to easily create cross reference between different document pages in way such that if refactor and reorganize docs/etc the links don't get broken. Curious in use use of backstage if this something other have had to tackle and if so best practices to do so.

Our team for now is using MyST and Sphinx to render full team docs from many repos into one. Though we are keeping eye on backstage as it's ability to integrate other things like allure, grads a, etc in one place would be awesome (and hoping that if success within our team we can promote around firm). For now given backstage is still bit newer we are keeping eye on project and feature to see if/when switch may make sense. Look forward to see continued progress in this project

**Comment by warsaw** [comment]
@garyniemen thanks for the update.  Sphinx and RST are still important to us.  Of course, I understand the Backstage project needs to set its own priorities, so this is something we are looking at building and possible/hopefully open sourcing if we come up with something reasonable.  We're looking at a number of approaches, including a generic plugin to surface GitHub Pages content in Backstage.

**Comment by garyniemen** [comment]
Thank you @ksdaftari and @warsaw - it was really interesting to read the experiences from your companies. We will keep these in mind as we discuss how we can more support more companies' documentation environments.
Just curious - for your current doc solutions - can you describe (just briefly) the documentation flow and where everything ends up. For example, do you integrate your Sphinx output with MyST output?

**Comment by warsaw** [comment]
Hi @garyniemen here's some background.

We currently have in production a somewhat bespoke doc publishing system, but it essentially boils down to using Sphinx to process docs for each commit to a product repo.  We don't use monorepo so every library, app, etc. lives in its own repo.  Once a merge to main/master occurs, the post-merge CI process runs a standard Sphinx build against the docs, uploading the resulting artifacts to our publishing system.  Most repos use `.rst` but some do use `.md`, although the latter have to declare additional dependencies on either `recommonmark` or `MyST-Parser` as the owners of those repos choose.  For this system, Sphinx is the only site generator option, but repo owners have control over their dependencies and `conf.py` file, so they have a lot of flexibility to choose styling and extensions within that constraint.

We're in the process of rolling out a new system where repo owners have more flexibility in the site generators they can choose, which of course comes with the possibility of less consistency across repos (using defaults and social pressure to encourage consistency).  Most pilots of the new system have chosen Docusaurus and `.md` files, which along with Sphinx are our two supported options (repo owners can use almost anything they want, but so far, none have chosen anything but one of these two defaults).  For `.md` files, our users are overwhelmingly happy with Docusaurus.  I'd say the majority of folks staying on `.rst` and thus Sphinx are those migrating from the old system, or Python developers for whom `.rst` often feels more natural, or has more features such as autogen of API docs, etc.  Sphinx has a rich ecosystem of plugins.

We've actually just recently completed an analysis of TechDocs for our use cases, and for now, we're not going to adopt it.  The main reasons are that many of our teams really want active React components in their docs, want to choose their own styling, or are using `.rst` files from the old system.  TechDocs has a different philosophy where it values consistency of docs as a major requirement -- which is perfectly reasonable of course! -- it's just not aligned with our own requirements.  So TechDocs stripping out JS/CSS and shadow-dom-ing the content into a Backstage card just doesn't work for us; too much useful stuff gets thrown away.  We built a PoC of a Backstage plugin which consumed built content from the doc repositories of our new system (GitHub Pages branches, but that's not super important), but iframe-ing doesn't work well for us either, for various reasons.

Put all that together and we're not able to utilize current TechDocs plugin for the repo-based docs, nor does it seem feasible to build our own plugin for Backstage.  That said, we have lots of opportunities for doc consolidation (e.g. product doc hubs, tech stack docs, and other high level entry points) and discovery, and those may be places where we can use TechDocs, essentially as landing pages which are consistent, easily discoverable, visible on the dev portal, and which link out to comprehensive repo-based docs.  So we'll be exploring that route too, while keeping an eye on TechDocs developments over the coming months and quarters.

Keep in mind that all of this documentation is for internal engineers, geared toward what they need to be efficient and effective in doing their jobs.  This is not documentation for external consumers.

I hope that gives you a high level sense of where we're going with docs, and how our interests intersect with TechDocs and documentation surfacing through Backstage.

**Comment by garyniemen** [comment]
@hexeberlin Sorry about slow response time. @warsaw Thank you for your detailed sharing of your context - totally understand. Here is our thinking...
The overall sentiment from the TechDocs maintainers is that we want to help all Backstage adopters use and be able to use TechDocs. To put it another way, we don't want not-being-able-to-use-TechDocs to be a blocker for adopting Backstage. Having said that, we have discussed in the team and here let me lay out what we see as the various options when it comes to making TechDocs work with other static site generators and other markup languages:
1. One way to solve this is for us to make TechDocs SSG-agnostic. We have discussed this a lot in the team and we are definitely tempted. But it is significant work and at this time this work is not prioritised.
2. There is the option for you to build your own Backstage plugin to address these use cases. However, you would probably need to replicate a lot of what we have in place already within TechDocs - for example, the CLI that can be used in CI/CD situations with nicely abstracted concepts (prepare/generate/publish), which also allow for just-in-time generation (no CI/CD necessary).
3. @hexeberlin in particular: Is Sphinx a must or is RST support what you are looking for? If the latter, one route could be to find a way to use techdocs-cli to support RST.
Here are some related issues:
https://github.com/backstage/backstage/issues/9172
https://github.com/backstage/backstage/issues/7048 (contains some further links)
